LocalPayment/src/main/java/com/braintreepayments/api/LocalPaymentClient.java

```diff
@@ -16,20 +18,74 @@
+    private LocalPaymentListener listener;
 
+    @VisibleForTesting
+    BrowserSwitchResult pendingBrowserSwitchResult;
+
+    /**
+     * Create a new instance of {@link LocalPaymentClient} from within an Activity using a {@link BraintreeClient}.
+     *
+     * @param activity a {@link FragmentActivity}
+     * @param braintreeClient a {@link BraintreeClient}
+     */
+    public LocalPaymentClient(@NonNull FragmentActivity activity, @NonNull BraintreeClient braintreeClient) {
+        this(activity, activity.getLifecycle(), braintreeClient, new PayPalDataCollector(), new LocalPaymentApi(braintreeClient));
+    }
+
+    /**
+     * Create a new instance of {@link LocalPaymentClient} from within a Fragment using a {@link BraintreeClient}.
+     *
+     * @param fragment a {@link Fragment
+     * @param braintreeClient a {@link BraintreeClient}
+     */
+    public LocalPaymentClient(@NonNull Fragment fragment, @NonNull BraintreeClient braintreeClient) {
+        this(fragment.getActivity(), fragment.getLifecycle(), braintreeClient, new PayPalDataCollector(), new LocalPaymentApi(braintreeClient));
+    }
+
+    /**
+     * Create a new instance of {@link LocalPaymentClient} using a {@link BraintreeClient}.
+     *
+     * Deprecated. Use {@link LocalPaymentClient(Fragment, BraintreeClient)} or
+     * {@link LocalPaymentClient(FragmentActivity, BraintreeClient)}.
+     *
+     * @param braintreeClient a {@link BraintreeClient}
+     */
+    @Deprecated
     public LocalPaymentClient(@NonNull BraintreeClient braintreeClient) {
-        this(braintreeClient, new PayPalDataCollector());
+        this(null, null, braintreeClient, new PayPalDataCollector(), new LocalPaymentApi(braintreeClient));
     }
 
     @VisibleForTesting
-    LocalPaymentClient(@NonNull BraintreeClient braintreeClient, @NonNull PayPalDataCollector payPalDataCollector) {
+    LocalPaymentClient(FragmentActivity activity, Lifecycle lifecycle, @NonNull BraintreeClient braintreeClient, @NonNull PayPalDataCollector payPalDataCollector, @NonNull LocalPaymentApi localPaymentApi) {
         this.braintreeClient = braintreeClient;
         this.payPalDataCollector = payPalDataCollector;
+        this.localPaymentApi = localPaymentApi;
+        if (activity != null && lifecycle != null) {
+            LocalPaymentLifecycleObserver observer = new LocalPaymentLifecycleObserver(this);
+            lifecycle.addObserver(observer);
+        }
+    }
+
+    /**
+     * Add a {@link LocalPaymentListener} to your client to receive results or errors from the Local Payment flow.
+     * This method must be invoked on a {@link LocalPaymentClient(Fragment, BraintreeClient)} or
+     * {@link LocalPaymentClient(FragmentActivity, BraintreeClient)} in order to receive results.
+     *
+     * @param context an Android Context
+     * @param listener a {@link LocalPaymentListener}
+     */
+    public void setListener(Context context, LocalPaymentListener listener) {
```

Comment:
This method currently requires context, because the `onBrowserSwitchResult` method requires context to call `PayPalDataCollector#getClientMetadataId(Context context)`. Looking at the docs for that method, it doesn't seem like we can pre-call this method, or cache the value.

Requiring context on a listener setter doesn't feel like a great merchant experience. This is only needed so that we don't fail silently when a merchant doesn't set their listener before trying to handle a browser switch result. We think in this case it would be better to remove `Context` from this method, and not try to deliver a pending browser switch result when the listener is set. We left `Context` in for now, so that you can see what the experience would look like. Open to other thoughts on the best approach for this.

Reply:
`BraintreeClient` has an accessor for `applicationContext` we could use that internally here.

Reply:
Ah of course - good call!